File: Cell_Phone_Prices/projects/Cell_Phone_Prices_1.py

import requests
import json as json
import numpy as np

import argparse
import tkinter as tk
import PySimpleGUI as sg

def get_multiple_predictions(values, fields):
    """
    This function calculates the predictions on
    multiple samples
    Args:
    values (dict): Dictionary containing url and feature values
    fields (list): List of strings containing feature names
    Returns:
    predictions (list): Contains the list of predictions
    """
    values_int_memory = values['int_memory']
    values_split = values_int_memory.split(',')

    arg_list = []

    for a in range(0,len(values_split)):
        args = {'url': values['url']}
        feature_num = 1
        body = '{'
        for feature in fields:
            if(feature == 'int_memory'):
                body = body + '"' + feature + '"' + ':' + str(values_split[a]) + ','
            else:
                body = body + '"' + feature + '"' + ':' + str(values[feature]) + ','
            feature_num = feature_num+1
        body = body[0:-1] + '}'
        args['body'] = body
        arg_list.append(args)
    predictions = []
    for queries in arg_list:
        prediction = json.loads(get_prediction(queries['url'],
                                               queries['body']))
        pred_value = json.loads(prediction['body'])['predicted_label']
        predictions.append(pred_value)
    return predictions


def ui_mode():
    """
    This function generates a UI where the user can enter
    feature values and see responses
    """
    # Create a dialog box with the features
    fields = ['battery_power','blue','clock_speed','dual_sim','fc','four_g','int_memory','m_dep','mobile_wt','n_cores',
              'pc','px_height','px_width','ram','sc_h','sc_w','talk_time','three_g','touch_screen','wifi']
    layout = [[sg.Text('url'), sg.Input('', key='url')],
              [sg.Text('battery_power'), sg.Input(803 , key='battery_power')],
              [sg.Text('blue'), sg.Input(1 , key='blue')],
              [sg.Text('clock_speed'), sg.Input(2.1, key='clock_speed')],
              [sg.Text('dual_sim'), sg.Input(0, key='dual_sim')],
              [sg.Text('fc'), sg.Input(7, key='fc')],
              [sg.Text('four_g'), sg.Input(0, key='four_g')],
              [sg.Text('int_memory'), sg.Input(17, key='int_memory')],
              [sg.Text('m_dep'), sg.Input(1.0, key='m_dep')],
              [sg.Text('mobile_wt'), sg.Input(198, key='mobile_wt')],
              [sg.Text('n_cores'), sg.Input(4, key='n_cores')],
              [sg.Text('pc'), sg.Input(11, key='pc')],
              [sg.Text('px_height'), sg.Input(344, key='px_height')],
              [sg.Text('px_width'), sg.Input(1440, key='px_width')],
              [sg.Text('ram'), sg.Input(2680, key='ram')],
              [sg.Text('sc_h'), sg.Input(7, key='sc_h')],
              [sg.Text('sc_w'), sg.Input(1, key='sc_w')],
              [sg.Text('talk_time'), sg.Input(4, key='talk_time')],
              [sg.Text('three_g'), sg.Input(1, key='three_g')],
              [sg.Text('touch_screen'), sg.Input(0, key='touch_screen')],
              [sg.Text('wifi', text_color='red'), sg.Input(1, key='wifi')],
              [sg.Text('Prediction: ', text_color='blue', size=(40, 2), key='_PRED_')],
              [sg.RButton('Predict'), sg.Exit()]]
    window = sg.Window('Enter url and feature values', layout)
    while True:
        event, values = window.Read()
        predictions = str(get_multiple_predictions(values, fields))
        predictions.replace('"', '')
        if event == 'Predict':
            print('entered predict event')
            # change the "output" element to be the value of "input" element
            window.Element('_PRED_').Update('int_memory: [' + values['int_memory'] + ']\n' +
                                            'Prediction: ' + predictions)
        if event is None or event == 'Exit':
            break

    window.Close()

def parseArguments():
    # Create argument parser
    parser = argparse.ArgumentParser()

    # Positional mandatory arguments
    parser.add_argument("--body", help="text in the format provided by snippet in Navigator", type=str)
    parser.add_argument("--url", help="Your endpoint URL", type=str)

    # Parse arguments
    try:
        args = parser.parse_args()
    except:
        return None

    return args


def get_prediction(url, data='{"battery_power" : 803,'
                              '"blue" : 1,'
                              '"clock_speed" : 2.1,'
                              '"dual_sim" : 0,'
                              '"fc" : 7,'
                              '"four_g" : 0,'
                              '"int_memory" : 17,'
                              '"m_dep" : 1.0,'
                              '"mobile_wt" : 198,'
                              '"n_cores" : 4,'
                              '"pc" : 11,'
                              '"px_height" : 344,'
                              '"px_width" : 1440,'
                              '"ram" : 2680,'
                              '"sc_h" : 7,'
                              '"sc_w" : 1,'
                              '"talk_time" : 4,'
                              '"three_g" : 1,'
                              '"touch_screen" : 0,'
                              '"wifi" : 1}'):
    r = requests.post(url, data=data)
    response = getattr(r, '_content').decode("utf-8")
    return response

if __name__ == '__main__':
    # Parse the arguments
    # Return None when no arguments are provided
    args = parseArguments()
    print('args: ', args)
    # Print the arguments provided, when present
    if args.url and args.body:
        for a in args.__dict__:
            print(str(a) + ": " + str(args.__dict__[a]))
        print(get_prediction(args.url, args.body))
    else:
        # When no arguments are provided, pop up a
        # dialog box where the user can enter
        # arguments
        ui_mode()

File: Cell_Phone_Prices/projects/CellPhonePrices1.java
import java.awt.*;
import java.awt.event.*;
import java.io.*;
import java.net.*;
import java.util.*;
import javax.swing.*;

public class CellPhonePrices1 extends JFrame implements ActionListener {

	private JPanel contentPane;
	private String[] labelNames = {"battery_power","blue","clock_speed","dual_sim","fc","four_g","int_memory","m_dep","mobile_wt","n_cores",
								"pc","px_height","px_width","ram","sc_h","sc_w","talk_time","three_g","touch_screen","wifi"};

	private String[] defaultValues = { "803", "1", "2.1","0","7","0","17","1.0","198","4","11","344","1440","2680","7","1","4","1","0","1"};

	private Map<String, JLabel> labels = new HashMap<>();
	private Map<String, JTextField> userInput = new HashMap<>();

	// x and y coordinates of panel, width and height of panel
	private int x, y, width, height;

	private String data;
	private JLabel prediction;
	private JTextField endpointURL;

	public CellPhonePrices1(int x, int y, int width, int height) {
		this.x = x;
		this.y = y;
		this.width = width;
		this.height = height;
		data = "";
		prediction = new JLabel("Prediction: ");
		endpointURL = new JTextField();
	}

	public String makePrediction() {
		URL url;
		String pred = "";
		try {
			url = new URL(endpointURL.getText());
			HttpURLConnection con = (HttpURLConnection) url.openConnection();
			con.setDoOutput(true);

			// write string for request
			OutputStream os = con.getOutputStream();
			byte[] input = data.getBytes("utf-8");
			os.write(input, 0, input.length);

			// read and output resulting prediction
			BufferedReader br = new BufferedReader(new InputStreamReader(con.getInputStream(), "utf-8"));
			StringBuilder response = new StringBuilder();
			String responseLine = null;
			while ((responseLine = br.readLine()) != null) {
				response.append(responseLine.trim());
			}

			// extracting prediction from response
			int index = response.indexOf("\\\"predicted_label\\\": ") + "\\\"predicted_label\\\": ".length();
			pred = response.substring(index);
			pred = pred.substring(0, pred.indexOf(","));

		} catch (IOException e) {
			e.printStackTrace();
		}

		return pred;
	}

	private void generatePanelContent() {

		// set panel configuration
		setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
		setBounds(this.x, this.y, this.width, this.height);
		setTitle("Click Through Rate Prediction Service");
		contentPane = new JPanel();
		setContentPane(contentPane);
		contentPane.setLayout(new GridLayout(labelNames.length + 2, 2));

		// added URL text field
		contentPane.add(new JLabel("URL"));
		contentPane.add(endpointURL);

		// adding attribute text fields to panel
		for (int i = 0; i < labelNames.length; i++) {
			labels.put(labelNames[i], new JLabel(labelNames[i]));
			userInput.put(labelNames[i] + "Input", new JTextField(defaultValues[i]));
			contentPane.add(labels.get(labelNames[i]));
			contentPane.add(userInput.get(labelNames[i] + "Input"));
		}

		JButton predictButton = new JButton("Predict");
		predictButton.addActionListener(this);
		contentPane.add(predictButton);
		contentPane.add(prediction);

	}

	@Override
	public void actionPerformed(ActionEvent e) {
		// getting data from text fields and making JSON
		data = "";
		data += "{";
		for (int i = 0; i < labelNames.length; i++) {
			data += "\"" + labelNames[i] + "\":\"" + userInput.get(labelNames[i] + "Input").getText() + "\"";
			if (i != labelNames.length - 1) {
				data += ",";
			}
		}
		data += "}";

		// gives user back prediction in GUI
		prediction.setText("Prediction: " + makePrediction());
	}

	public static void main(String[] args) {
		CellPhonePrices1 price_range = new CellPhonePrices1(0, 0, 400, 400);
		price_range.generatePanelContent();
		price_range.setVisible(true);
	}

}
